### DISCUSSION TITLE
Pickling A Stumpi Object

### DISCUSSION THREAD

**Discussion by rnjv** [TDAmeritrade/stumpy#486] [Help: Other Questions]
Hi,

I am trying to pickle the Stumpi class but I get the following error:
PicklingError: Can't pickle <class 'stumpy.stumpi.stumpi'>: it's not the same object as stumpy.stumpi.stumpi

How do I save and load Stumpi to disk?

```
    mp = stumpy.stumpi(T, m, egress=False)

    with open('cache/interim/mp/{}_stumpi.pkl'.format(id), 'wb') as f:
        pickle.dump(mp, f)
```

The above I am running from a Jupyter notebook.

**Comment by seanlaw** [comment]
@rnjv Unfortunately, I have little to no experience with pickling.

@hmcoservit I vaguely remember that you had tried pickling `aampi` [here](https://github.com/TDAmeritrade/stumpy/issues/342#issuecomment-785101747). Do you happen to remember what you had to do in order to Pickle a STUMPY object?

**Reply by seanlaw** [reply]
@rnjv I think [this](https://stackoverflow.com/questions/34253069/how-to-use-class-decorators-with-pickle) is related to your problem. Essentially, behind the scenes, `stumpy.stumpi` is wrapped by a decorator and it's interfering with your pickling. I'd have to play around and think about whether there is an easy solution to work around this.

**Reply by seanlaw** [reply]
Okay, so it may not be the answer that you were hoping for but if you are able to use `cloudpickle` (an arguably better alternative to the built-in `pickle`) then you can do:

```
#!/usr/bin/env python

import numpy as np
import stumpy
import cloudpickle


if __name__ == "__main__":
    T = np.random.rand(100)
    m = 50

    stream = stumpy.stumpi(T, m, egress=True)
    stream.update(np.random.rand())
    print(stream.left_P_)

    dump_stream = cloudpickle.dumps(stream)

    stream = cloudpickle.loads(dump_stream)
    print(stream.left_P_)
```

I hope that helps!

**Reply by hmcoservit** [reply]
Hi @seanlaw, sorry I it was a long time ago I no longer have access to that code.
